pybamm/discretisations/base_discretisation.py

```diff
@@ -0,0 +1,306 @@
+#
+# Interface for discretisation
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pybamm
+
+import numpy as np
+
+
+class BaseDiscretisation(object):
+    """The base discretisation class, with methods to process a model and replace
+    Spatial Operators with Matrices and Variables with StateVectors
+
+    Parameters
+    ----------
+    mesh : :class:`BaseMesh` (or subclass)
+        The underlying mesh for discretisation
+
+    """
+
+    def __init__(self, mesh):
+        self._mesh = mesh
+
+    @property
+    def mesh(self):
+        return self._mesh
+
+    def process_model(self, model):
+        """Discretise a model.
+        Currently inplace, could be changed to return a new model.
+
+        Parameters
+        ----------
+        model : dict
+            Model ({variable: equation} dict) to dicretise
+
+        Returns
+        -------
+        y0 : :class:`numpy.array`
+            Vector of initial conditions
+        dydt : method
+            A method that takes a time t and array y and returns the an array of
+            concatenated time-derivatives.
+
+        """
+        # Set the y split for variables
+        y_slices = self.get_variable_slices(model.rhs.keys())
+
+        # Discretise and concatenate initial conditions, passing domain from variable
+        y0 = self.process_initial_conditions(model.initial_conditions)
+
+        # Discretise right-hand sides, passing domain from variable
+        dydt = self.process_rhs(model.rhs, model.boundary_conditions, y_slices)
+
+        return y0, dydt
+
+    def get_variable_slices(self, variables):
+        """Set the slicing for variables.
+
+        Parameters
+        ----------
+        variables : dict_keys object containing Variable instances
+            Variables for which to set the slices
+
+        Returns
+        -------
+        y_slices : dict of {variable id: slice}
+            The slices to take when solving (assigning chunks of y to each vector)
+        """
+        y_slices = {variable.id: None for variable in variables}
+        start = 0
+        end = 0
+        for variable in variables:
+            # Add up the size of all the domains in variable.domain
+            for dom in variable.domain:
+                end += getattr(self.mesh, dom).npts
+            y_slices[variable.id] = slice(start, end)
```

Comment:
I think something like this is necessary because nodes are being copied when added to the expression tree, so a `Variable` in the rhs equation won't automatically recognise that it is the same as the `Variable` in the keys. Implementing `__hash__` and `__eq__` for symbols isn't possible because it raises loop errors in the tree (`anytree` thinks that the same variable is being put in twice)

Reply:
I see, that is a bit of a pain :( Should we do the same thing for the `rhs` dict in `Model` then? I.e. make the keys the ids?

Reply:
I think that would expose a bit of ugliness to someone who just wants to add new models

Reply:
ok, we'll just leave it alone then.